### PR TITLE
Pre commit without cargo and other pre-PR improvements

### PR DESCRIPTION
**Pull request by konstin**
This tackles three problems:
* pre-commit was slow because it ran cargo commands
* Improve the clarity on what you need to run to get your PR pass on CI (and make those fast)
* You had to compile and run `cargo dev generate-all` separately, which was slow

The first change is to remove all cargo commands except running ruff itself from pre-commit. With `cargo run --bin ruff` already compiled it takes about 7s on my machine. It would make sense to also use the ruff pre-commit action here even if we're then lagging a release behind for checking ruff on ruff.

The contributing guide is now clear about what you need to run:

```shell
cargo clippy --workspace --all-targets --all-features -- -D warnings  # Linting...
RUFF_UPDATE_SCHEMA=1 cargo test  # Testing and updating ruff.schema.json
pre-commit run --all-files  # rust and python formatting, markdown and python linting, etc.
```

Example timings from my machine:

`cargo clippy --workspace --all-targets --all-features -- -D warnings`: 23s
`RUFF_UPDATE_SCHEMA=1 cargo test`: 2min (recompiling), 1min (no code changes, this is mainly doc tests)
`pre-commit run --all-files`: 7s

The exact numbers don't matter so much as the approximate experience (6s is easier to just wait than 1min, esp if you need to fix and rerun). The biggest remaining block seems to be doc tests, i'm surprised i didn't find any solution to speeding them up (nextest simply doesn't run them at all). Also note that the formatter has it's own tests which are much faster since they avoid linking ruff (`cargo test ruff_python_formatter`).

The third change is to enable `cargo test` to update the schema. Similar to `INSTA_UPDATE=always`, i've added `RUFF_UPDATE_SCHEMA=1` (name open to bikeshedding), so `RUFF_UPDATE_SCHEMA=1 cargo test` updates the schema, while `cargo test` still fails as expected if the repo isn't up-to-date.

### CODE REVIEW

**Comment by konstin** [comment]
the benchmarks are really noisy, there are no code changes in the linter or the formatter

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -57,9 +63,9 @@ Prior to opening a pull request, ensure that your code has been auto-formatted,
 and that it passes both the lint and test validation checks:
 
 ```shell
-cargo fmt  # Auto-formatting...
-cargo test  # Testing...
 cargo clippy --workspace --all-targets --all-features -- -D warnings  # Linting...
+RUFF_UPDATE_SCHEMA=1 cargo test  # Testing and updating ruff.schema.json
+pre-commit run --all-files  # rust and python formatting, markdown and python linting, etc.
```
Comment (dhruvmanila) [inline CONTRIBUTING.md:68]: I always use this flag to see the actual changes made by the hooks. Do you think it's useful?

```suggestion
pre-commit run --all-files --show-diff-on-failure  # rust and python formatting, markdown and python linting, etc.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -57,9 +63,9 @@ Prior to opening a pull request, ensure that your code has been auto-formatted,
 and that it passes both the lint and test validation checks:
 
 ```shell
-cargo fmt  # Auto-formatting...
-cargo test  # Testing...
```
Comment (charliermarsh) [inline CONTRIBUTING.md:61]: Why remove formatting and testing here?

File: crates/ruff_dev/src/generate_json_schema.rs [hunk 3]
```diff
@@ -50,13 +55,19 @@ pub(crate) fn main(args: &Args) -> Result<()> {
...
+    use std::env;
 
     use crate::generate_all::Mode;
 
     use super::{main, Args};
 
     #[test]
     fn test_generate_json_schema() -> Result<()> {
-        main(&Args { mode: Mode::Check })
+        let mode = if env::var("RUFF_UPDATE_SCHEMA") == Ok("1".to_string()) {
```
Comment (charliermarsh) [inline crates/ruff_dev/src/generate_json_schema.rs:66]: Nit: `env::var("RUFF_UPDATE_SCHEMA").as_ref() == Ok("1")` (no need to allocate)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -45,6 +45,12 @@ You'll also need [Insta](https://insta.rs/docs/) to update snapshot tests:
 cargo install cargo-insta
 ```
 
+and pre-commit to run some further linter:
```
Comment (charliermarsh) [inline CONTRIBUTING.md:48]: Maybe: "And [`pre-commit` ](https://pre-commit.com/#installation), to run some validation checks:"

**Review by charliermarsh — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (konstin) [reply CONTRIBUTING.md:61]: formatting is done by pre-commit, testing just moved below clippy

File: CONTRIBUTING.md (hunk 1 above)
Comment (konstin) [reply CONTRIBUTING.md]: this is neat, thanks

**Pull request merged**